src/connection.c

```diff
@@ -345,6 +345,16 @@ static int connSocketGetType(connection *conn) {
     return CONN_TYPE_SOCKET;
 }
 
+static int connSocketLastErrorRetryable(connection *conn) {
+    if (conn->last_errno == EINTR) {
+        conn->state = CONN_STATE_CONNECTED;
+        conn->last_errno = 0;
```

Comment:
i don't think this method should take any action (change state or reset the errno).
i think it should just return an indication.
i.e. `return conn->last_errno == EINTR;`

I think the code that got the error in the first place (and returned -1), should have had the check to avoid setting the state to FAIL (i.e. like the previous commit did).

or am i'm misinterpreting what @yossigo intended?

Reply:
At present, we don't want to change the connection state to `CONN_STATE_ERROR` when some errors occur (At present, the errors include EAGAIN and EINTR). Should we add a new interface connSetRetryAbleError (int errno) to register retryable errors? Like this:
```c
conn = tls ? connCreateTLS() : connCreateSocket();
connSetRetryAbleError(conn, EINTR);
```
in connection.c, if it is a retryable error, `CONN_STATE_ERROR` is not set, but set last_errno, like this:
```c
static int connSocketRead(connection *conn, void *buf, size_t buf_len) {
    int ret = read(conn->fd, buf, buf_len);
    if (!ret) {
        conn->state = CONN_STATE_CLOSED;
    } else if (ret < 0) {
        conn->last_errno = errno;
        if (!isRetryableError(errno) && conn->state == CONN_STATE_CONNECTED)
            conn->state = CONN_STATE_ERROR;
    }
    return ret;
}
```
in rio.c, If connRead() returns -1, like thie:
```c
int retval = connRead(conn);
if (retval <=0) {
    if (connLastErrorRetryable(r->io.conn.conn) == C_OK) continue;
    /* use connGetLastError(conn) log. */
    return 0;
}
```
@oranagra @yossigo WDYT?

Reply:
i'm not sure i follow your last comment.
i think the code in connRead in this commit https://github.com/menwenjun/redis/commit/020afce7acbd52617a2448947a3ea3684ad82518 looks decent, in the fact it avoids setting the state of the connection to failed, but what it didn't do correctly is lose the value of last_errno.

the current code in the PR is worse since the getter function (connLastErrorRetryable) clears that, so calling it twice, the second time would return the wrong value.

i.e. the key points are:
1. connRead will return -1 on any failure
2. last_error will always be set
3. it'll avoid setting the to CONN_STATE_ERROR if last_error is EAGAIN
4. both connGetLastError and connLastErrorRetryable can be called repeatedly (they don't clear any state variable).

am i missing anything?

Reply:
`so calling it twice, the second time would return the wrong value.` Sorry, I ignored that. I revised the code again, please CR.

Reply:
C code LGTM.